### PR TITLE
text.usetex ticklabel alignment problem solution

### PR DESCRIPTION
**Pull request by avinashmnit30**
## PR Summary
This text.usetex ticklabel alignment problem mentioned in #11243 . The code uses amsmath negative sign when` 'text.latex.preamble' : [r'\usepackage{amsmath}'] `is updated in `rcParams`. 
`

### CODE REVIEW

**Comment by ImportanceOfBeingErnest** [comment]
I suppose this would fail in case you use something like `\usepackage{amsmath, amssymb, amsthm}` in the preamble.

Also it seems to only affect cases, where `set_xticklabels` is used, it would hence only affect a very small subset of all plots.

**Comment by avinashmnit30** [comment]
You are right, but it would work if i use `'amsmath' in rcParams['text.latex.preamble'][0]):` instead of `r'\usepackage{amsmath}' in rcParams['text.latex.preamble']):`

**Comment by ImportanceOfBeingErnest** [comment]
Well that produces an error in case no preamble is defined. 
More generally, it would also fail in case the letters `amsmath` appear in the preamble without that package actually be used.

**Comment by avinashmnit30** [comment]
ticklabels = get_labels

        get_labels = []
        if(rcParams['text.usetex'] and
            len(rcParams['text.latex.preamble']) != 0):
            for presemble in rcParams['text.latex.preamble']:
                if('usepackage{'  in presemble and 'amsmath' in presemble):
                    # check text.usetex and amsmath package 
                    # if True use amsmath package for the negative sign
                    for t in ticklabels:
                        # replace latex "-" sign with latex.amsmath "-" sign
                        get_labels.append(t.replace('-','\\text{-}'))
                    ticklabels = get_labels
                    break

**Comment by avinashmnit30** [comment]
`len(rcParams['text.latex.preamble']) != 0):` would solve the undefined preamble error. 
`if('usepackage{'  in presemble and 'amsmath' in presemble):` would solve the 2nd problem when amsmath appaer in the package without being used.
  
```
        if(rcParams['text.usetex'] and
            len(rcParams['text.latex.preamble']) != 0):
            for presemble in rcParams['text.latex.preamble']:
                if('usepackage{'  in presemble and 'amsmath' in presemble):
                    # check text.usetex and amsmath package 
                    # if True use amsmath package for the negative sign
                    for t in ticklabels:
                        # replace latex "-" sign with latex.amsmath "-" sign
                        get_labels.append(t.replace('-','\\text{-}'))
                    ticklabels = get_labels
                    break
```

**Comment by ImportanceOfBeingErnest** [comment]
As [the documentation](https://matplotlib.org/tutorials/text/usetex.html?highlight=preamble#troubleshooting) states
>The text.latex.preamble rc setting is not officially supported.

so it feels a bit strange to then support it in one single function for ticklabels.  
Again, this fix would only work for the specific case where ticklabels are set via `set_ticklabels` which is only one of the many possible ways to produce ticklabels in general.

Concerning the actual code, as said, it is not sufficient to just check if the letters `amsmath` are in the preamble. Consider e.g.

```
rctex = {"text.usetex" : True,
         "text.latex.preamble" : r"\usepackage{xcolor} % dont use amsmath here"}
plt.rcParams.update(rctex)
```
where `amsmath` is in the preamble, but it's not used, because it's only a comment.

**Comment by anntzer** [comment]
It would be nice to know *why* this fixes the issue.

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1643,6 +1643,19 @@ def set_ticklabels(self, ticklabels, *args, minor=False, **kwargs):
...
+        get_labels = []
+        if(rcParams['text.usetex'] and
+            len(rcParams['text.latex.preamble']) != 0):
+            for presemble in rcParams['text.latex.preamble']:
+                if('usepackage{'  in presemble and 'amsmath' in presemble):
+                    # check text.usetex and amsmath package 
+                    # if True use amsmath package for the negative sign
+                    for t in ticklabels:
+                        # replace latex "-" sign with latex.amsmath "-" sign
+                        get_labels.append(t.replace('-','\\text{-}'))
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: Would mathrm work?

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1643,6 +1643,19 @@ def set_ticklabels(self, ticklabels, *args, minor=False, **kwargs):
         # replace the ticklabels list with the processed one
         ticklabels = get_labels
 
+        get_labels = []
+        if(rcParams['text.usetex'] and
```
Comment (anntzer) [inline lib/matplotlib/axis.py]: Remember that text objects can each have their individual usetex setting, the global setting is just a default.  So it's the text's usetex value that needs to be checked.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (avinashmnit30) [reply lib/matplotlib/axis.py]: `\mathrm{-} `does not work. I tried. But maybe `\operatorname{-}`. In that case **presemble** check won't be required.

Comment (avinashmnit30) [reply lib/matplotlib/axis.py]: \mathrm{-}does not work. I tried. But maybe \operatorname{-}. In that case presemble check won't be required.

**Comment by avinashmnit30** [comment]
@anntzer This is a latex rendering issue not directly related to matplotlib. It seems that latex renders `\\text{-}` and `-`  signs differently. And you right we can use `\operatorname{-}` instead of `\text{-}`. Now I have updated it at draw function with text object's usetex setting.

**Comment by ImportanceOfBeingErnest** [comment]
Now I'm confused. The current version checks the preamble for amsmath, while `\operatorname`  does not require amsmath, right? But then again the example from above with something like `r"\usepackage{xcolor} % dont use amsmath here"` in the preamble would make this fail.

**Pull request closed without merging**